src/write_gcm.c

```diff
@@ -502,29 +502,29 @@ static EVP_PKEY *wg_credential_contex_load_pkey(char const *filename,
 //==============================================================================
 //==============================================================================
 
+typedef struct {
+  char *data;
+  size_t size;
+} wg_curl_write_ctx_t;
+
 // Does an HTTP GET or POST, with optional HTTP headers. The type of request is
 // determined by 'body': if 'body' is NULL, does a GET, otherwise does a POST.
 // If curl_easy_init() or curl_easy_perform() fail, returns -1.
 // If they succeed but the HTTP response code is >= 400, returns -2.
 // Otherwise returns 0.
-static int wg_curl_get_or_post(char *response_buffer,
-    size_t response_buffer_size, const char *url, const char *body,
-    const char **headers, int num_headers, _Bool silent_failures);
+static int wg_curl_get_or_post(wg_curl_write_ctx_t *ctx, const char *url,
```

Comment:
There's no real reason to pass in the context -- you'll just have to initialize it to the same value in every caller (`.data = NULL, .size = 0`). I would pass in a pointer to the response buffer pointer (i.e., `char **response_ptr`), only use `wg_curl_write_ctx_t` inside this function, and then assign `*response_ptr = write_ctx.data` (we don't really care about `size` for the NUL-terminated string anyway).

Reply:
The only reason I'll rebut this is the point that @kosak made before, we can use the context to signal a true allocation error vs the data just being null in general. It allows us to pass around more meaningful signals than just using the response buffer pointer. WDYT?

Reply:
(Again I'm feeling guilty about parachuting into this conversation). I think @igorpeshansky's suggestion is nice, and it can still preserve the concept of buffer reuse if you want it to. The semantics would be that the caller passes in NULL if there is not an existing allocation, otherwise the caller passes in the existing buffer.  (And, thanks to the way ```realloc``` works, this would not require any new 'if' statements or special cases)

the code would be like

```
char *response = NULL;
while (list != NULL) {
    ...
    int wg_result = wg_curl_get_or_post(&response, url, body, headers...)) {
...
}

leave:
   sfree(response);
```

The point is that wg_curl_get_or_post will reuse the ```response``` buffer if it's non-null.

Reply:
Sorry, only one comment from me, feel free to ignore it if you think it doesn't apply. I know this is C, but generally it's an anti-pattern to allocate memory in one "module" and release in another. Because the allocators aren't remembered, you could end up the wrong allocator.

Even if today wg_curl_get_or_post is defined in the same file as its callers, I could easily see that changing if we want to reuse that function.

Reply:
@jkohen Are we doing this anti-pattern in either case whether we pass in a struct or a char*? I just want to be sure whether or not you feel the current implementation is also an "anti-pattern", and how I could go about allocating memory up front.